src/typecheck.rs

```diff
@@ -17,9 +17,32 @@
 //!
 //! # Type inference
 //!
-//! Type inference is done via a standard unification algorithm. The type of unannotated let-bound
-//! expressions (the type of `bound_exp` in `let x = bound_exp in body`) is inferred in strict
-//! mode, but it is never implicitly generalized. For example, the following program is rejected:
+//! ## Algorithm
+//!
+//! Type inference is done via a mix of unification and bidirectional typechecking. Basically, an
+//! annotation triggers the traditional checking mode of bidirectional typechecking. When no
+//! checking rule applies anymore, and we need inference, the typechecker just spins up a
+//! unification variable and check against this type. At such interfaces between checking and
+//! inference, we take the occasion to apply a subtyping constraint, that is to check that
+//! `Inferred <: InitiallyChecked`, as customary in such systems (see [Bidirectional typing, Jana
+//! Dunfield and Neel Krishnaswami](https://arxiv.org/abs/1908.05839) for an good survey of
+//! bidirectional typing).
+//!
+//! Subtyping checks are implemented in `check_sub` and similar sub-functions. We don't perform any
+//! elaborate inference with respect to subtyping.  Solving subtyping equations is known to be hard
+//! and to result in more complex types being inferred. Instead, we decompose the constraints until
+//! we reach one that has a plain unification variable on one side.  Then, any such constraint of
+//! the form `a <: Type` or `Type <: a` results in the standard unification of `a` and `Type`,
+//! instead of recording e.g. lower and upper bounds and solving inequalities later. Doing so,
+//! inference is simpler and more predictable while retaining most of the practical convenience of
+//! subtyping, which is about being able to convert to a less precise type seamlessly (`Num` to
+//! `Dyn, or `{a : Num}` to `{_ : Num}`).
```

Comment:
I think that the LHS of a subtyping constraint is not necessarily an inferred type, for example in a situation like `let x : T1 = exp in (exp :  T2)` or `let x : T1 = .. in let f : T2 -> T3 = .. in f x`, that would give rise to constraints `T1 <: T2` for user-written (hence potentially polymorphic) types `T1` and `T2`.

> I realise that something something deep instantiation; but as argued in A quick look at impredicativity you probably don't want this anyway

I'm not sure to parse this: what is this that we don't want anyway? Deep instantiation? Those kind of left foralls rules, even if those look necessary precisely to avoid deep instantiation? Sorry, I haven't read the QuickLook paper yet, just looked at the talk.

Concerning the simplify subsumption proposal, @goldfirere did tell me about GHC dropping e.g. the arrow rule, and that this was linked to impredicativity. However DK's system is predicative, and I think we probably never gonna need nor want impredicativity in Nickel. I'm gonna take a look at the proposal, but are there other reasons than impredicativity to drop the arrow and deep skolemization rules?

Coming back to QuickLook, once again I don't think we want or need impredicativity. That said, as opposed to the vanilla DK system, we have more subtyping going on than just polymorphism, so we do have the problem of the order of arguments affecting the instantiation:

```
let fst: forall a. a -> a -> a = fun x y => x in
let x = {foo = 1, bar = 2} in
let y : {_ : Num} = {baz = 3} in

fst x y // Fail, and will still fail even with full DK
fst y x // Succeed
```

I guess we could apply exactly QuickLook's idea (from my superficial understanding) of "use all the arguments at a call site to try to decide the most general instantiation for a polymorphic type" in the same way, indeed.

Reply:
> we probably never gonna need nor want impredicativity in Nickel

This sounds like something you will regret someday! :)

>  guess we could apply exactly QuickLook's idea (from my superficial understanding) of "use all the arguments at a call site to try to decide the most general instantiation for a polymorphic type" in the same way, indeed.

Yes. That example is a good one, and it does suggest to me that you want the QL algorithm in some form. I think order-independence among function arguments is important.

To illustrate why QL wants to get rid of the arrow rule (just using subtyping w.r.t. polymorphism, not Nickel's full subtyping):
1. Suppose we are checking `(forall a. a -> a) <:? ((forall b. b) -> Int)`.
2. We will instantiate the `a` with a unification variable `alpha`: `(alpha -> alpha) <:? ((forall b. b) -> Int)`
3. Now, if we have the arrow rule, we decompose: `(forall b. b) <:? alpha` and `alpha <:? Int`.
4. Now there are two solutions: `alpha := forall b. b` and `alpha := Int`. Both will work here, but one might be better than the other in a more elaborate circumstance.

On the other hand, if we required `(forall b. b) =? alpha` (that is, we didn't have the co-/contra-variant arrow rule), then there would be only one solution. The arrow rule gives us too much flexibility, and this flexibility can make type inference unpredictable and order-dependent. This is why QL jettisoned the arrow rule.

In Nickel, I see similar situations arising: because of your expressive subtype relation, switching to equality when looking under an arrow will allow for fewer -- and thus more predictable -- choices.